coveo-testing/README.md

```diff
@@ -221,3 +221,456 @@ If you run `pytest --collect-only` you will obtain the following:
+
+```python
+from unittest.mock import patch, MagicMock
+from coveo_testing.mocks import ref
+
+from mymodule.tasks import process
+from mymodule.clients import _get_api_client
+
+
+@patch(*ref(_get_api_client))
+def test(api_client_mock: MagicMock) -> None:
+    assert process()
+```
+
+
+### Mock something for a given context
+
+When you want to mock something for a given module, you must provide a hint to `ref` as the `context` argument.
+
+The hint may be a module, or a function/class defined within that module. "Defined" here means that "def" or "class" 
+was used _in that module_. If the hint was imported into the module, it will not work:
+
+`mymodule.tasks`:
+
+```python
+from mymodule.clients import get_api_client
+
+def process() -> bool:
+  client = get_api_client()
+  return ...
+```
+
+The test, showing 3 different methods that work:
+
+```python
+from unittest.mock import patch, MagicMock
+from coveo_testing.mocks import ref
+
+from mymodule.clients import get_api_client
+from mymodule.tasks import process
+
+# you can pass the module as the context
+import mymodule
+
+@patch(*ref(get_api_client, context=mymodule.tasks))
+def test(get_api_client_mock: MagicMock) -> None:
+    assert process()
+
+# you can pass the module as the context, version 2
+from mymodule import tasks
+    
+@patch(*ref(get_api_client, context=tasks))
+def test(get_api_client_mock: MagicMock) -> None:
+    assert process()
+
+# you can also pass a function or a class defined in the `tasks` module
+from mymodule.tasks import process
+@patch(*ref(get_api_client, context=process))
+def test(get_api_client_mock: MagicMock) -> None:
+    assert process()
+```
+
+The 3rd method is encouraged: provide the function or class that is actually using the `get_api_client` import.
+In our example, that's the `process` function.
+If `process` was ever moved to a different module, it would carry the `get_api_client` import, and the mock would
+be automatically adjusted to target `process`'s new module without changes.
+
+### Mock something for the current context
+
+Sometimes, the test file _is_ the context. When that happens, just pass `__name__` as the context:
+
+```python
+from unittest.mock import patch
+from coveo_testing.mocks import ref
+from mymodule.clients import get_api_client, APIClient
+
+def _prepare_test() -> APIClient:
+    client = get_api_client()
+    ...
+    return client
+    
+@patch(*ref(get_api_client, context=__name__))
+def test() -> None:
+    client = _prepare_test()
+    ...
+```
+
+### Mock a function or class at the root level
+
+When mocking at the root level, you need to be careful about the context:
+
+```python
+with patch(*ref(my_function)): ...
+with patch(*ref(my_function, context=...)): ...
+with patch(*ref(MyClass)): ...
+with patch(*ref(MyClass, context=...)): ...
+```
+
+The `context` is a slightly confusing topic that is explained in the `Demo` above.

```

Comment:
I'm not sure that I understand what's being explained here. I think that I'm confused because I don't quite get what you mean by "root level" in this context. Do you mean when using `with patch(...):` outside of a function at the root of the module?

Perhaps the explanation should go deeper here. I read this as: "Be careful & check out the demo". My reaction to reading this is: "Be careful of what? What could go wrong?".

Reply:
i removed that part, it was redundant.